src/main/kotlin/uk/gov/dluhc/printapi/service/PrintResponseProcessingService.kt

```diff
@@ -0,0 +1,92 @@
+     * The reason for the new requestId is, the print provider does not support duplicate print requests with the same requestId.
+     *
+     * dateCreated will be the timestamp the new printRequestStatus is created and eventDateTime will be the timestamp from
+     * the print provider's batch response.
+     */
+    fun processBatchResponses(batchResponses: List<BatchResponse>) {
+        batchResponses.forEach { batchResponse ->
+            with(batchResponse) {
+                val newStatus =
+                    if (status == SUCCESS) RECEIVED_BY_PRINT_PROVIDER else PENDING_ASSIGNMENT_TO_BATCH
+
+                val printDetails =
+                    printDetailsRepository.getAllByStatusAndBatchId(SENT_TO_PRINT_PROVIDER, batchId)
+
+                printDetails.forEach {
+                    it.addStatus(
+                        status = newStatus,
+                        dateCreated = OffsetDateTime.now(clock),
+                        eventDateTime = timestamp,
+                        message = null
+                    )
+
+                    if (status != SUCCESS) {
+                        it.requestId = idFactory.requestId()
+                    }
+                }
+
+                printDetailsRepository.updateItems(printDetails)
+            }
+        }
+    }
+
+    fun processPrintResponse(printResponse: ProcessPrintResponseMessage) {
+        val newStatus = with(printResponse) {
+            when {
+                statusStep == ResponseStatusStep.PROCESSED && status == ResponseStatus.SUCCESS -> VALIDATED_BY_PRINT_PROVIDER
+                statusStep == ResponseStatusStep.IN_MINUS_PRODUCTION && status == ResponseStatus.SUCCESS -> IN_PRODUCTION
+                statusStep == ResponseStatusStep.DISPATCHED && status == ResponseStatus.SUCCESS -> DISPATCHED
+                statusStep == ResponseStatusStep.NOT_MINUS_DELIVERED && status == ResponseStatus.FAILED -> NOT_DELIVERED
+                statusStep == ResponseStatusStep.PROCESSED && status == ResponseStatus.FAILED -> PRINT_PROVIDER_VALIDATION_FAILED
+                statusStep == ResponseStatusStep.IN_MINUS_PRODUCTION && status == ResponseStatus.FAILED -> PRINT_PROVIDER_PRODUCTION_FAILED
+                statusStep == ResponseStatusStep.DISPATCHED && status == ResponseStatus.FAILED -> PRINT_PROVIDER_DISPATCH_FAILED
+                else -> throw IllegalArgumentException("Undefined statusStep [$statusStep] and status [$status] combination")
+            }
+        }
+
+        val printDetails = printDetailsRepository.getByRequestId(printResponse.requestId)
+
+        with(printResponse) {
+            printDetails.addStatus(
+                status = newStatus,
+                dateCreated = OffsetDateTime.now(clock),
+                eventDateTime = timestamp,
+                message = message
+            )
+        }
+
+        printDetailsRepository.updateItems(listOf(printDetails))
```

Comment:
Is there a better method on the repository (could there be?)
It seems a bit weird to call a method that is about updating many items, when we only have one 🤔

Reply:
@NathanRussellValtech , this is the only update method in the dynamodb repository and since we will be switching to mysql shortly, I didn't want to introduce a new method to the repository at this point.